lsp-mode.el

```diff
@@ -4185,7 +4185,13 @@ Return a nested alist keyed by symbol names. e.g.
 
 (defun lsp-enable-imenu ()
   "Use lsp-imenu for the current buffer."
-  (setq-local imenu-create-index-function #'lsp--imenu-create-index))
+  (setq-local imenu-create-index-function #'lsp--imenu-create-index)
+  (imenu--cleanup)
+  ;; Make sure imenu-update-menubar redoes everything.
+  (setq imenu-menubar-modified-tick -1)
+  (setq imenu--index-alist nil)
+  (setq imenu--last-menubar-index-alist nil)
+  (imenu-update-menubar))
```

Comment:
I am not very familiar with the imenu related code: is this needed when you do not have menubar enabled? If not, I would prefer skipping this call and let imenu update the cache when imenu is called explicitly.

Reply:
Thank you for the prompt reply!

The key thing to note is that there is a delay in the reporting od the client's Imenu capability.  Therefore, `imenu-create-index-function` is not set "early enough," so to say.

Since Imenu updatea its alist under certain conditions, the current code can result in Imenu using a "pre-LSP" alist.

This happens in two scenarios:

1. Imenu has a menu-bar entry, which will compute the alist very likely before LSP does anything since it runs on `menu-bar-update-hook`, or
2. The user forces the alist to be computed by invoking `imenu` while `lsp-mode` is waiting for capability response.

This basically clears any alist that may have been computed before the LSP client indicated its Imenu capability.  Lastly, `imenu-update-menubar` does nothing if there is no menu-bar entry for Imenu.